packages/react/src/hooks/useEditorState.ts

```diff
@@ -0,0 +1,229 @@
+import type { BlockNoteEditor } from "@blocknote/core";
+import deepEqual from "fast-deep-equal/es6/react";
+import { useDebugValue, useEffect, useLayoutEffect, useState } from "react";
+import { useSyncExternalStoreWithSelector } from "use-sync-external-store/shim/with-selector";
+import { useBlockNoteContext } from "../editor/BlockNoteContext.js";
+
+const useIsomorphicLayoutEffect =
+  typeof window !== "undefined" ? useLayoutEffect : useEffect;
+
+export type EditorStateSnapshot<
+  TEditor extends BlockNoteEditor<any, any, any> | null = BlockNoteEditor<
+    any,
+    any,
+    any
+  > | null,
+> = {
+  editor: TEditor;
+  transactionNumber: number;
+};
+
+export type UseEditorStateOptions<
+  TSelectorResult,
+  TEditor extends BlockNoteEditor<any, any, any> | null = BlockNoteEditor<
+    any,
+    any,
+    any
+  > | null,
+> = {
+  /**
+   * The editor instance. If not provided, will use the editor from BlockNoteContext.
+   */
+  editor?: TEditor;
+
+  /**
+   * A selector function to determine the value to compare for re-rendering.
+   */
+  selector: (context: EditorStateSnapshot<TEditor>) => TSelectorResult;
+
+  /**
+   * A custom equality function to determine if the editor should re-render.
+   * @default `deepEqual` from `fast-deep-equal`
+   */
+  equalityFn?: (a: TSelectorResult, b: TSelectorResult | null) => boolean;
+
+  /**
+   * The event to subscribe to.
+   * @default "all"
+   */
+  on?: "all" | "selection" | "change";
+};
+
+/**
+ * To synchronize the editor instance with the component state,
+ * we need to create a separate instance that is not affected by the component re-renders.
+ */
+class EditorStateManager<
```

Comment:
I suppose this is mostly copied from tiptap? What were the changes needed?

Reply:
The only change needed here was the `on` and being able to pass in the BlockNoteEditor instance and typing for that to work properly. Otherwise is what I wrote for Tiptap